### ISSUE TITLE
Websocket spamming too much data

### ISSUE THREAD

**Issue by svk31** [cryptonomex/graphene#540]
The witness_node websocket connection is sending out ~200KB of data on every block, that's 4MB per minute, and this is without counting any other objects that are being sent out, this is only the constant data being sent with every new block.

The reason is that the bitasset_data_id objects for mosts assets are included in every frame. Inspecting those objects, most of them have no changes, while a few have very tiny changes (like 3s increment to current_feed_publication_time). If they aren't being changed, why are they included??

**Comment by theoreticalbts** [comment]
I think I know why.  I'll release a patch shortly.

**Comment by theoreticalbts** [comment]
Try recompiling with this patch (you'll have to merge it into your BitShares checkout and recompile).

**Comment by svk31** [comment]
I tried this but it didn't make any difference to the objects being pushed.

**Comment by kenCode-de** [comment]
Thanx for working on this guys :)

**Comment by kenCode-de** [comment]
Hey guys just checking in on this one. Just fyi, our mobile wallet will hit the first public beta by friday of next week, would be great if I didn't have to warn people first about it's data usage. Please and thanx again :)

**Comment by dnotestein** [comment]
this issue is a good reason to vote for our worker (Blockchain maintenance developer), it's first on the list of issues we plan to tackle after we get voted in

**Comment by abitmore** [comment]
> Hey guys just checking in on this one. Just fyi, our mobile wallet will hit the first public beta by friday of next week, would be great if I didn't have to warn people first about it's data usage. Please and thanx again :)

Even if it's fixed, it won't be so soon to be deployed imo. However a hard fork is not required, so it depends on when OpenLedger will deploy it. Hopefully sooner :)

**Comment by theoreticalbts** [comment]
Since @dnotestein seems to want to tackle this one, I'm assigning it to him.

**Comment by kenCode-de** [comment]
Hey, if this solves the problem of data consumption over mobile, then I am stoked. Please and thanx @dnotestein and crew :)

**Comment by kenCode-de** [comment]
Will I have to upgrade anything in our POS or mobile wallets once this bug is fixed? Our stuff just sits on top of your stuff and we are just talkin to a Witness so our stuff should immediately leverage the new reduced bandwidth utilization automatically, right?

**Comment by dnotestein** [comment]
yes, that's correct

**Comment by kenCode-de** [comment]
this is great news, thank you guys! :)

**Comment by kenCode-de** [comment]
thanx for that commit @emfrias I hope that, along with the OP here, will finally drop data and cpu utilization way back to help speed up our mobile apps. Thank you :)

**Comment by theoreticalbts** [comment]
Fix #614 doesn't work as-is, and if it is modified to work correctly, won't fix this bug.

`feed_is_expired()` returns the opposite value of what's expected, but [the caller](https://github.com/cryptonomex/graphene/blob/e3478d233847056943da6f84f18adcf18c42e64d/libraries/chain/db_update.cpp#L437-L443) takes that into account and does the thing that should be done if it is not expired when the `feed_is_expired()` return value is `true`.

So the proper fix would also have the caller check `!feed_is_expired()` instead of `feed_is_expired()` as currently, which would end up not changing any semantics.

What this fix effectively does (without the call site change) is mark all BitAssets as expired and stop updating the median feed.  So I'm sure it will help with the websocket spamming problem, but it will also break all BitAssets!

**Comment by theoreticalbts** [comment]
In #615 @abitmore asks a ~~provocative~~ thought-provoking question:

> Why did you think (according to the code) when no feed is expired, we should still change the settle price in db?

The reason is that there is no easy way to tell whether the feed is expired with the data structures and semantics we have.  We store the feed entries in a `flat_map` in the `asset_bitasset_data_object` and can only index them by `account_id_type`.  So the only way to determine "is there any expired?" is to walk through all of them.

The walking logic is implemented as `asset_bitasset_data_object::update_median_feeds(time_point_sec)`.  Doing the querying (determining whether the object needs to be modified) and updating (actually modifying the object) in the same method seems to be an optimization (otherwise you'd have to walk twice, once to determine whether modification is necessary and again to perform the modification).

After analyzing the code, I think the correct fix is to write the output of `update_median_feeds` into an intermediate object allocated on the caller's stack, then copy this intermediate object's fields to the database iff there was a change.

**Comment by abitmore** [comment]
@theoreticalbts Firstly, I apologize if I've ever been offensive.

> The reason is that there is no easy way to tell whether the feed is expired with the data structures and semantics we have. We store the feed entries in a flat_map in the asset_bitasset_data_object and can only index them by account_id_type. So the only way to determine "is there any expired?" is to walk through all of them.

Secondly, IMHO, the easy way to tell whether the feed is expired (so need to be calculated again) is to set the oldest unexpired feed publish time (which is published by witnesses and stored in the `flat_map`) as "current" feed publish time, which is done with [this code](https://github.com/cryptonomex/graphene/blob/e3478d233847056943da6f84f18adcf18c42e64d/libraries/chain/asset_object.cpp#L56), then check it every block, as we already did [here](https://github.com/cryptonomex/graphene/blob/e3478d233847056943da6f84f18adcf18c42e64d/libraries/chain/db_update.cpp#L437-L443) and [here](https://github.com/cryptonomex/graphene/blob/e3478d233847056943da6f84f18adcf18c42e64d/libraries/chain/include/graphene/chain/asset_object.hpp#L212-L215). When the oldest feed expires, recalculate the median price of the asset, and update the publish time of "current feed" price time to next oldest unexpired one at same time (which have been done in `update_median_feeds`).

> So the only way to determine "is there any expired?" is to walk through all of them.
> Actually the problem is that update_median_feeds() is called every block, which unconditionally modifies the asset_bitasset_data_object for every MIA.

Thirdly, if we need to call `update_median_feeds()` every block, why do we bother doing [this check](https://github.com/cryptonomex/graphene/blob/e3478d233847056943da6f84f18adcf18c42e64d/libraries/chain/db_update.cpp#L437) in the code?

**Comment by theoreticalbts** [comment]
@abitmore : I wasn't offended, I used the wrong word (edited).  You've been a very helpful contributor lately and I appreciate your efforts.

I think I misinterpreted both your question and the code -- I thought you were asking why we change the settle price in DB when no feed is expired, and then when I read the code, I just totally glossed over the `if( is_feed_expired() )`.

Actually this has led me to reconsider #615 ( and your patch #614 ).  On second thought you were right -- the logic is supposed to only update the feed when it expires, instead it is updating when the feed is non-expired, leading to many repetitions of objects (and non-updating when feeds expire).  Moving discussion to #615 (largely because this is getting really long and there's an unrelated branch numbered 540, which was my initial fix attempt).
